### PR TITLE
Issue #14891: MissingSwitchDefault should not violate switch that con…

### LINKED ISSUE #14891: Add Support for Java 21 Record Pattern : MissingSwitchDefault
child of #14890 

I have read check documentation: https://checkstyle.org/checks/coding/missingswitchdefault.html#MissingSwitchDefault
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---

```java

PS D:\CS\test> javac src/RecordPatterns.java --enable-preview --source 22
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.List;

public class RecordPatterns {
    sealed interface I permits C, D { }
    final class C implements I { }
    final class D implements I { }
    record Pair<T>(T x, T y) { }

    void test() {
        C c = new C();
        D d = new D();

        Pair<I> p2 = new Pair<>(c, d);
        // exhaustiveness check by the compiler
        switch (p2) {   // violation ??
            case Pair<I>(I i, C j) -> System.out.println("Pair (C|D, C)");
            case Pair<I>(I i, D j) -> System.out.println("Pair (C|D, D)");
        }
    }

    public static void main(String[] args) {
        new RecordPatterns().test();
    }
}
PS D:\CS\test> cat config.xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="MissingSwitchDefault"/>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:15:9: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 1 errors.

``` 

**Describe what you expect in detail.**

I expect no violation here since there is an exhaustiveness Check by the compiler. This means that all possible inputs must be covered. So there is no need for default

---

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #14891 :

Diff Regression config: https://gist.githubusercontent.com/mahfouz72/afd57aa6e51ca161b7927acd959e5211/raw/fd6bb80bfe66feb1d7bf946caf87c0a7be981f6c/check.xml

Diff Regression projects: https://gist.githubusercontent.com/mahfouz72/a3d0af030c8f5efd0d8a39f2c14750bc/raw/a3424ad9b6f722e5de1a927e3d85e383101b2b93/projects-to-test-on.properties

### CODE REVIEW

**Comment by mahfouz72** [comment]
Github, generate report

**Comment by romani** [comment]
@nrmancuso , @rnveach , try think of we should collect such test bundles (config , list of projects) in our test-configs repository.
This specific is simple default config, but what if any non default come up let's seriously consider to preserve it, or signal to me and Piyush to review.

@relentless-pursuit , as you finish generation of basic set of configs, you will be immediately huge help for others. People will not use gist anymore for simple cases.

**Review by nrmancuso — APPROVED** [review]
Let’s go!

**Review by rnveach — COMMENTED** [review]
Add one more input example where the switch does have a default.

https://checkstyle.org/checks/coding/missingswitchdefault.html#Description
> This check does not validate switch statements that use pattern or null labels. Rationale: Switch statements that use pattern or null labels are checked by the compiler for exhaustiveness.

You made a point to update the javadoc. Shouldn't this be updated as well?

> See the Java Language Specification for more information about switch statements and expressions

Shouldn't we update this to the new JLS also show the new patterns?

**Comment by mahfouz72** [comment]
> > This check does not validate switch statements that use pattern or null labels. Rationale: Switch statements that use pattern or null labels are checked by the compiler for exhaustiveness.
> 
> You made a point to update the javadoc. Shouldn't this be updated as well?

I assumed that the word "pattern" in "that use pattern or null labels." includes both normal type patterns and record pattern

**Comment by mahfouz72** [comment]
Github, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java [hunk 1]
```diff
@@ -55,6 +55,10 @@
  *     and expressions.
  * </p>
  * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30">
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java:58]: > new record pattern syntax introduced in Java 21

I would expect use to use the new JLS in which it was introduced. Update the one before it as well.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java (hunk 1 above)
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MissingSwitchDefaultCheck.java]: done

**Comment by mahfouz72** [comment]
Github, generate site

**Review by rnveach — APPROVED** [review]

**Pull request merged**